cinn/ir/ir_compare.cc

```diff
@@ -0,0 +1,308 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cinn/ir/ir_compare.h"
+
+#include "cinn/ir/ir_base.h"
+#include "cinn/ir/ir_printer.h"
+
+namespace cinn {
+namespace ir {
+
+bool IrEqualVisitor::Compare(const Expr& lhs, const Expr& rhs) {
+  if (lhs.get() == rhs.get()) {  // the same object, including both are null
+    return true;
+  }
+
+  bool equal = lhs.defined() && rhs.defined();  // someone invalid
+  equal      = equal && lhs->node_type() == rhs->node_type();
+  equal      = equal && IRVisitorBase<bool, const Expr*>::Visit(&lhs, &rhs);
+
+  if (!equal) {
+    VLOG(5) << "Not equal on Expr, lhs:[type:" << kIrNodeTyReprs[static_cast<int>(lhs->node_type())] << "]\n"
+            << lhs << ", \nrhs[type:" << kIrNodeTyReprs[static_cast<int>(rhs->node_type())] << "]\n"
+            << rhs;
+  }
+  return equal;
+}
+
+bool IrEqualVisitor::Compare(const std::string& lhs, const std::string& rhs, bool allow_name_suffix_diff) {
+  // if allow_name_suffix_diff=true then just compare the name prefix before the "_[0-9]+"
+  auto remove_idx_fn = [](const std::string& name) -> std::string {
+    auto idx_pos = name.find_last_of('_');
+    return idx_pos == std::string::npos ? name : name.substr(0, idx_pos);
+  };
+
+  bool equal = true;
+  if (!allow_name_suffix_diff) {
+    equal = lhs == rhs;
+  } else {
+    auto lhs_prefix = remove_idx_fn(lhs);
+    auto rhs_prefix = remove_idx_fn(rhs);
+    equal           = lhs_prefix == rhs_prefix;
+  }
+
+  if (!equal) {
+    VLOG(5) << "Not euqal on name, lhs=" << lhs << ", rhs=" << rhs;
+  }
+
+  return equal;
+}
+
+bool IrEqualVisitor::Compare(const std::map<std::string, attr_t>& lhs, const std::map<std::string, attr_t>& rhs) {
+  if (lhs.size() != rhs.size()) {
+    VLOG(6) << "Not equal on attrs, lhs size=" << lhs.size() << ", rhs size=" << rhs.size();
+    return false;
+  }
+  for (auto&& kv : lhs) {
+    auto opposite = rhs.find(kv.first);
+    if (opposite == rhs.end() || kv.second != opposite->second) {
+      VLOG(6) << "Not equal at attr key=" << kv.first;
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename T>
+bool IrEqualVisitor::Compare(const std::vector<T>& lhs, const std::vector<T>& rhs) {
+  if (lhs.size() != rhs.size()) {
+    VLOG(6) << "Not equal on repeated fields, lhs size=" << lhs.size() << ", rhs size=" << rhs.size();
+    return false;
+  }
+  for (auto i = 0; i < lhs.size(); ++i) {
+    if (!Compare(lhs.at(i), rhs.at(i))) {
+      VLOG(6) << "Not equal on repeated fields at index=" << i;
+      return false;
+    }
+  }
+  return true;
+}
+
+#define PRIMITIVE_TYPE_IMPL(op__)                                  \
+  bool IrEqualVisitor::Visit(const op__* lhs, const Expr* other) { \
+    auto* rhs = other->As<op__>();                                 \
```

Comment:
Should we check whether `rhs` is `nullptr` in these visit functions? `As` returns `nullptr` when `Expr` is not the `As` type.

Reply:
No, because it has been checked on the top `Compare`  and only when both lhs and rhs are defined, these `Visit` will be called by `IRVisitorBase<bool, const Expr*>::Visit(&lhs, &rhs)`

Reply:
Got you